src/qibolab/instruments/dummy.py

```diff
@@ -55,38 +56,38 @@ def stop(self):
     def disconnect(self):
         log.info("Disconnecting dummy instrument.")
 
-    def get_values(self, options, sequence, exp_points):
+    def get_values(self, options, sequence, shape):
         results = {}
         for ro_pulse in sequence.ro_pulses:
             if options.acquisition_type is AcquisitionType.DISCRIMINATION:
                 if options.averaging_mode is AveragingMode.SINGLESHOT:
-                    values = np.random.randint(2, size=exp_points)
+                    values = np.random.randint(2, size=shape)
                 elif options.averaging_mode is AveragingMode.CYCLIC:
-                    values = np.random.rand(exp_points)
+                    values = np.random.rand(*shape)
             elif options.acquisition_type is AcquisitionType.RAW:
                 samples = int(ro_pulse.duration * self.sampling_rate)
-                values = np.random.rand(samples * exp_points) * 100 + 1j * np.random.rand(samples * exp_points) * 100
+                waveform_shape = tuple(samples * dim for dim in shape)
+                values = np.random.rand(*waveform_shape) * 100 + 1j * np.random.rand(*waveform_shape) * 100
             elif options.acquisition_type is AcquisitionType.INTEGRATION:
-                values = np.random.rand(exp_points) * 100 + 1j * np.random.rand(exp_points) * 100
+                values = np.random.rand(*shape) * 100 + 1j * np.random.rand(*shape) * 100
             results[ro_pulse.qubit] = results[ro_pulse.serial] = options.results_type(values)
         return results
 
-    def play(self, qubits: Dict[Union[str, int], Qubit], sequence: PulseSequence, options: ExecutionParameters):
+    def play(self, qubits: Dict[QubitId, Qubit], sequence: PulseSequence, options: ExecutionParameters):
         exp_points = 1 if options.averaging_mode is AveragingMode.CYCLIC else options.nshots
-
-        return self.get_values(options, sequence, exp_points)
+        shape = (exp_points,)
+        return self.get_values(options, sequence, shape)
 
     def sweep(
         self,
-        qubits: Dict[Union[str, int], Qubit],
+        qubits: Dict[QubitId, Qubit],
         sequence: PulseSequence,
         options: ExecutionParameters,
         *sweepers: List[Sweeper],
     ):
-        exp_points = 1
-        for sweeper in sweepers:
-            exp_points *= len(sweeper.values)
         if options.averaging_mode is not AveragingMode.CYCLIC:
-            exp_points *= options.nshots
-
-        return self.get_values(options, sequence, exp_points)
+            nshots = (options.nshots,)
+            shape = nshots + tuple(len(sweeper.values) for sweeper in sweepers)
```

Comment:
Not important, but I would also consider
```suggestion
            shape = (options.nshots,) + tuple(len(sweeper.values) for sweeper in sweepers)
```
to avoid setting `nshots` to a `tuple`, because usually when I see `nshots` around Qibo I expect `int`.